### PR TITLE
(feat) Ingest configurations should be committed gracefully

### PR DESCRIPTION
**Pull request by sivakuna-aap**

### CODE REVIEW

**Comment by amagdas** [comment]
:+1:

File: superdesk/io/commands/update_ingest.py [hunk 1]
```diff
@@ -77,29 +77,30 @@ class UpdateIngest(superdesk.Command):
     def run(self, provider_type=None):
         for provider in superdesk.get_resource_service('ingest_providers').get(req=None, lookup={}):
             if is_valid_type(provider, provider_type) and is_scheduled(provider) and not is_closed(provider):
-                update_provider.delay(str(provider['_id']))
+                rule_set = None
+                if provider.get('rule_set'):
+                    rule_set = superdesk.get_resource_service('rule_sets').find_one(_id=provider['rule_set'], req=None)
 
+                update_provider.delay(provider, rule_set)
```
Comment (ioanpocol) [inline superdesk/io/commands/update_ingest.py]: Provider and rule_set are serialized on update_provider.delay call so probably is better to send provider_id and read the required data on update_provider.

Comment (sivakuna-aap) [inline superdesk/io/commands/update_ingest.py]: If `update_provider.delay()` fails to execute immediately then we will not meet the requirement. That's the reason we are serializing the data. Do you foresee any problem with serialization?

Comment (ioanpocol) [inline superdesk/io/commands/update_ingest.py]: OK, so the intention is that any change to provider config and rule_set will not be applied to tasks already on processing.

File: superdesk/io/commands/update_ingest.py [hunk 2]
```diff
@@ -77,29 +77,30 @@ class UpdateIngest(superdesk.Command):
...
+    updates the provider.
+    """
 
     for items in providers[provider.get('type')].update(provider):
-        ingest_items(provider, items)
+        ingest_items(items, provider, rule_set)
 
     superdesk.get_resource_service('ingest_providers').update(provider['_id'], {
+        LAST_UPDATED: utcnow(),
         'config': provider.get('config', {}),  # persist changes to config if any
```
Comment (ioanpocol) [inline superdesk/io/commands/update_ingest.py]: Why is this needed? 
There is the case that an old task will overwrite the provider config that has been changed after the old task has been started. 
Also ETAG is changed only on patch so not sure if it is needed to be sent on update. It can be the same situation to overwrite a newer set ETAG.

Comment (petrjasek) [inline superdesk/io/commands/update_ingest.py]: re etag: idea is not to change etag via system, so if user will start editing a provider and it would run ingest update meanwhile you won't be able to save changes.

File: superdesk/io/commands/update_ingest.py [hunk 3]
```diff
@@ -77,29 +77,30 @@ class UpdateIngest(superdesk.Command):
...
+    Fetches items from ingest provider as per the configuration, ingests them into Superdesk and
+    updates the provider.
+    """
 
     for items in providers[provider.get('type')].update(provider):
-        ingest_items(provider, items)
+        ingest_items(items, provider, rule_set)
 
     superdesk.get_resource_service('ingest_providers').update(provider['_id'], {
+        LAST_UPDATED: utcnow(),
```
Comment (petrjasek) [inline superdesk/io/commands/update_ingest.py]: I think last updated should not be modified after but rather before because it's used for schedule and if the task will take longer it will run more of them. That's why there were 2 updates, not sure if we need the second one (for persisting changes to config)

File: superdesk/io/commands/update_ingest.py (hunk 2 above)
Comment (ioanpocol) [inline superdesk/io/commands/update_ingest.py]: Yes, but ETAG is changed on patch() not on update(), so by calling update() the tag will not be changed.

Comment (petrjasek) [inline superdesk/io/commands/update_ingest.py]: it will be changed if not provided

File: superdesk/io/commands/update_ingest.py (hunk 1 above)
Comment (sivakuna-aap) [inline superdesk/io/commands/update_ingest.py]: Yes, any changes to either config or rule_sets shouldn't be applied to the tasks in progress.

File: superdesk/io/commands/update_ingest.py (hunk 3 above)
Comment (sivakuna-aap) [inline superdesk/io/commands/update_ingest.py]: Agreed and will restore the first updated but I think we don't need the second update.

**Comment by amagdas** [comment]
:+1:

**Comment by petrjasek** [comment]
:+1:

**Comment by petrjasek** [comment]
I'm just wondering if we need to deal with changes to config between the time task is scheduled and executed. It seems idea is not to change config for an update that is in progress already right? I would say scheduled is not yet in progress and as long as it fetched the config only once while running it should be fine.

**Comment by amagdas** [comment]
Needs rebase.

**Comment by ioanpocol** [comment]
:+1:

**Comment by mdhaman** [comment]
Ignore this PR. Instead use #483

**Comment by amagdas** [comment]
Closing then...
